plugins/inputs/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,690 @@
+package stackdriver
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/api/iterator"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal/limiter"
+	"github.com/influxdata/telegraf/plugins/inputs"
+
+	// Imports the Stackdriver Monitoring client package.
+	monitoring "cloud.google.com/go/monitoring/apiv3"
+	googlepbduration "github.com/golang/protobuf/ptypes/duration"
+	googlepbts "github.com/golang/protobuf/ptypes/timestamp"
+	distributionpb "google.golang.org/genproto/googleapis/api/distribution"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+)
+
+const (
+	description  = "Plugin that scrapes Google's v3 monitoring API."
+	sampleConfig = `
+## GCP Project (required - must be prefixed with "projects/")
+project = "projects/{project_id_or_number}"
+
+## API rate limit. On a default project, it seems that a single user can make
+## ~14 requests per second. This might be configurable. Each API request can
+## fetch every time series for a single metric type, though -- this is plenty
+## fast for scraping all the builtin metric types (and even a handful of
+## custom ones) every 60s.
+# rateLimit = 14
+
+## Collection Delay Seconds (required - must account for metrics availability via Stackdriver Monitoring API)
+# delaySeconds = 60
+
+## The first query to stackdriver queries for data points that have timestamp t
+## such that: (now() - delaySeconds - lookbackSeconds) <= t <= (now() - delaySeconds).
+## The subsequence queries to stackdriver query for data points that have timestamp t
+## such that: lastQueryEndTime <= t <= (now() - delaySeconds).
+## Note that influx will de-dedupe points that are pulled twice,
+## so it's best to be safe here, just in case it takes GCP awhile
+## to get around to recording the data you seek.
+# lookbackSeconds = 120
+
+## Metric collection period
+interval = "1m"
+
+## Configure the TTL for the internal cache of timeseries requests.
+## Defaults to 1 hr if not specified
```

Comment:
I don't think this actually defaults to 1 hour. I could be missing it, but I don't see any code that sets that default.

The function newTimeSeriesConf should be called per-timeseries every interval if there is no cache, but only once per cacheTTLSeconds if there is a cache. I added a debugging statement to this function

When I  specified an interval of 1m and did not set cacheTTLSeconds, my debugging statement appeared once per timeseries per minute. That shows there was actually no hour-long cache of the metric names. Metric data was output properly.

When I specified an interval of 1m and set `cacheTTLSeconds = 3600`, my debugging statment ran once per timeseries and then did not run again in the following minutes. That suggests the cache is there now. However, the plugin stopped outputting any metrics from stackdriver!

Reply:
Actually, I take back the portion about enabling a cache causing stackdriver metric output to stop. Something else has gone wrong, because now with cache commented out again my earlier change isn't working anymore.